test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.cc

```diff
@@ -0,0 +1,315 @@
+// TODO(ikepolinsky): Major action items to improve this fuzzer
+// 1. Move external process from separate thread to have test all in one thread
+//    - Explore using fake gRPC client for this
+// 2. Implement sending trailers from downstream and mutating headers/trailers
+//    in the external process.
+// 3. Use an upstream that sends varying responses (also with trailers)
+// 4. Explore performance optimizations:
+//    - Threads and fake gRPC client above might help
+//    - Local testing had almost 800k inline 8 bit counters resulting in ~3
+//      exec/s. How far can we reduce the number of counters?
+//    - At the loss of reproducibility use a persistent envoy
+// 5. Protobuf fuzzing would greatly increase crash test case readability
+//    - How will this impact speed?
+//    - Can it be done on single thread as well?
+
+#include "envoy/config/core/v3/base.pb.h"
+#include "envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.pb.h"
+#include "envoy/service/ext_proc/v3alpha/external_processor.pb.h"
+#include "envoy/type/v3/http_status.pb.h"
+
+#include "source/common/network/address_impl.h"
+
+#include "test/common/http/common.h"
+#include "test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz_helper.h"
+#include "test/extensions/filters/http/ext_proc/test_processor.h"
+#include "test/fuzz/fuzz_runner.h"
+#include "test/integration/http_integration.h"
+#include "test/test_common/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ExternalProcessing {
+
+using envoy::extensions::filters::http::ext_proc::v3alpha::ProcessingMode;
+using envoy::service::ext_proc::v3alpha::ProcessingRequest;
+using envoy::service::ext_proc::v3alpha::ProcessingResponse;
+
+// The buffer size for the listeners
+static const uint32_t BufferSize = 100000;
+
+// These tests exercise the ext_proc filter through Envoy's integration test
+// environment by configuring an instance of the Envoy server and driving it
+// through the mock network stack.
+
+class ExtProcIntegrationFuzz : public HttpIntegrationTest,
+                               public Grpc::BaseGrpcClientIntegrationParamTest {
+public:
+  ExtProcIntegrationFuzz(Network::Address::IpVersion ip_version, Grpc::ClientType client_type)
+      : HttpIntegrationTest(Http::CodecType::HTTP2, ip_version) {
+    ip_version_ = ip_version;
+    client_type_ = client_type;
+  }
+
+  void tearDown() {
+    cleanupUpstreamAndDownstream();
+    test_processor_.shutdown();
+  }
+
+  Network::Address::IpVersion ipVersion() const override { return ip_version_; }
+  Grpc::ClientType clientType() const override { return client_type_; }
```

Comment:
Out of curiosity, some of these methods seem like they would be the same as in base classes, do they need overriding?

Reply:
Since I inherit from `BaseGrpcClientIntegrationParamTest I did have to implement them. Will address context of why I inherit from `BaseGrpcClientIntegrationParamTest` in the comment on `ExtProcIntegrationTest` below.